File: java-springboot-toy/src/main/java/com/ex/javaspringboottoy/repository/QueryDslRepositorySupportCustom.java
package com.ex.javaspringboottoy.repository;

import com.querydsl.core.types.Expression;
import com.querydsl.core.types.dsl.Wildcard;
import com.querydsl.jpa.impl.JPAQuery;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.support.QuerydslRepositorySupport;

import java.util.List;
import java.util.Objects;

public class QueryDslRepositorySupportCustom extends QuerydslRepositorySupport {

    public QueryDslRepositorySupportCustom(Class<?> domainClass) {
        super(domainClass);
    }

    public <T> Page<T> getPage(JPAQuery<Object> query, Expression<T> selectClause, Pageable pageable) {
        Long totalCount = query.select(Wildcard.count).fetchFirst();
        List<T> content = Objects.requireNonNull(getQuerydsl()).applyPagination(pageable, query.select(selectClause)).fetch();
        return new PageImpl(content, pageable, totalCount);
    }
}

File: spring-hibernate/self-join-query/src/main/kotlin/com/toy/selfjoinquery/repository/QuerydslRepositorySupportCustom.kt
package com.toy.selfjoinquery.repository

import com.querydsl.core.types.Expression
import com.querydsl.core.types.dsl.Wildcard
import com.querydsl.jpa.impl.JPAQuery
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.support.QuerydslRepositorySupport

class QuerydslRepositorySupportCustom(domainClass: Class<*>) : QuerydslRepositorySupport(domainClass) {

  fun <T> getPage(
    query: JPAQuery<out Any>,
    selectClause: Expression<T>,
    pageable: Pageable
  ): PageImpl<T> {
    val totalCount = query.select(Wildcard.count).fetchFirst()
    val list = querydsl!!.applyPagination(pageable, query.select(selectClause)).fetch()

    return PageImpl(list, pageable, totalCount)
  }

}